tests/DCPS/Thrasher/DataReaderListenerImpl.cpp

```diff
@@ -1,78 +1,87 @@
+  : mutex_()
+#ifdef ACE_HAS_CPP11
+  , condition_()
+#else
+  , condition_(mutex_)
+#endif
+  , expected_samples_(expected_samples)
+  , received_samples_(0)
+  , task_samples_map_()
+  , progress_(progress_fmt, expected_samples_)
 {}
 
 DataReaderListenerImpl::~DataReaderListenerImpl()
 {}
 
+void DataReaderListenerImpl::wait_received()
+{
+  Lock lock(mutex_);
+  ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) sub wait_received %d:%d\n"), received_samples_, expected_samples_));
+  while (received_samples_ < expected_samples_) {
+#ifdef ACE_HAS_CPP11
+    condition_.wait(lock);
+#else
+    condition_.wait();
+#endif
+    ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) sub condition_.wait returned\n")));
+  }
+}
+
+int DataReaderListenerImpl::check_received(const size_t n_publishers) const
+{
+  ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) sub check_received\n")));
+  int ret = 0;
+  const size_t samples_per_publisher = expected_samples_ / n_publishers;
+  for (size_t p = 0; p < n_publishers; ++p) {
+    TaskSamplesMap::const_iterator i = task_samples_map_.find(p);
+    if (i != task_samples_map_.end()) {
+      for (size_t s = 0; s < samples_per_publisher; ++s) {
+        if (i->second.count(s) == 0) {
+          ACE_DEBUG((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: missing pub%d sample%d\n"), p, s));
+          ++ret;
+        }
+      }
+    } else {
+      ACE_DEBUG((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: missing pub%d samples\n"), p));
+      ++ret;
+    }
+  }
+  if (received_samples_ != expected_samples_) {
+    ACE_DEBUG((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: sub received %d of expected %d samples.\n"),
+      received_samples_, expected_samples_));
+    ++ret;
+  }
+  return ret;
+}
+
 void DataReaderListenerImpl::on_data_available(DDS::DataReader_ptr reader)
 {
   FooDataReader_var reader_i = FooDataReader::_narrow(reader);
-  if (CORBA::is_nil(reader_i)) {
-    ACE_ERROR((LM_ERROR,
-               ACE_TEXT("%N:%l: on_data_available()")
-               ACE_TEXT(" _narrow failed!\n")));
+  if (!reader_i) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("%N:%l: on_data_available() _narrow failed!\n")));
     return;
   }
 
-  // The following is intentionally inefficient to simulate
-  // backpressure with multiple writers; we should be especially
-  // dull and take only one sample at a time.
-
+  // Intentionally inefficient to simulate backpressure with multiple writers:
+  // take only one sample at a time.
   Foo foo;
   DDS::SampleInfo si;
-
+  bool less_than_expected = true;
   while (reader_i->take_next_sample(foo, si) == DDS::RETCODE_OK) {
     if (si.valid_data) {
-      ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+      Lock lock(mutex_);
       ++received_samples_;
       ++progress_;
-      task_sample_set_map[(size_t) foo.x].insert((size_t) foo.y);
-      condition_.broadcast();
+      task_samples_map_[(size_t) foo.x].insert((size_t) foo.y);
+      less_than_expected = received_samples_ < expected_samples_;
+    }
+    if (si.valid_data && !less_than_expected) {
```

Comment:
There are actually three things to consider here:
1) to access received_samples_ when the lock is held.
2) to notify the waiting thread when the condition is met, i.e. all the expected_samples_ have been received.
3) to notify the waiting thread when the lock is not held, to avoid the scenario that the notified thread would immediately block again, waiting for the notifying thread to release the lock.

Given the constraint that the std objects (mutex, condition_variable, unique_lock) and the predicated overload of wait(), wait_for(), or wait_until() are not always available across the configurations we support, implementing the above 1) is straight forward, but not 2) and 3).

Code1:
    if (si.valid_data) {
      Lock lock(mutex_);
      ++received_samples_;
      ++progress_;
      task_samples_map_[(size_t) foo.x].insert((size_t) foo.y);
      if (received_samples_ == expected_samples_) {
        condition_.notify_all(); //?? notify under lock
      }
    }

Code2:
    if (si.valid_data) {
      {
        Lock lock(mutex_);
        ++received_samples_;
        ++progress_;
        task_samples_map_[(size_t) foo.x].insert((size_t) foo.y);
      }
      if (received_samples_ == expected_samples_) { //?? access received_samples_ without lock
        condition_.notify_all();
      }
    }

That's why a local variable less_than_expected is used. It's updated under lock by checking received_samples_, and checked without lock so as to notify the waiting thread. The current implementation traded checking si.valid_data twice for the nested scope {}. If this is not a test, the nested scope {} should probably be used.

Reply:
It's fine (preferable, in fact) to notify under the lock. The other threads still must acquire the murex held by this thread before they are allowed to continue, which means your scenario 3 is impossible.

Reply:
std::condition_variable::notify_all
Notes
"The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock."

The above quote can be demonstrated by adding 3 seconds sleep after notification to exaggerate the situation.

1. notify without holding the lock

    if (si.valid_data) {
      {
        Lock lock(mutex_);
        ++received_samples_;
        ++progress_;
        task_samples_map_[(size_t) foo.x].insert((size_t) foo.y);
        received_all = received_samples_ >= expected_samples_;
      }
      if (received_all) {
        ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t)  sub condition_.notify_all\n")));
        condition_.notify_all();
        ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t)  sub sleep 3 seconds;\n")));
        ACE_OS::sleep(ACE_Time_Value(3, 0));
        ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t)  sub sleep returned\n")));
      }
    }
/*
(10969|123145466273792)  sub 100% (1024 samples received)
(10969|123145466273792)  sub condition_.notify_all //?? notify without the lock
(10969|123145466273792)  sub sleep 3 seconds;
(10969|4721579456) sub condition_.wait returned //?? waiting thread got the lock immediately
(10969|4721579456) <- PublisherService::end
(10969|123145498468352) pub58<-delete_participant
(10969|123145466273792)  sub sleep returned //?? notifying thread continues
(10969|123145493639168) pub49<-delete_participant
*/

2. notify while holding the lock

    if (si.valid_data) {
      Lock lock(mutex_);
      ++received_samples_;
      ++progress_;
      task_samples_map_[(size_t) foo.x].insert((size_t) foo.y);
      if (received_samples_ >= expected_samples_) {
        ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t)  sub condition_.notify_all under lock\n")));
        condition_.notify_all();
        ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t)  sub sleep 3 seconds;\n")));
        ACE_OS::sleep(ACE_Time_Value(3, 0));
        ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t)  sub sleep returned\n")));
      }
    }
/*
(11151|123145408430080)  sub 100% (1024 samples received)
(11151|123145408430080)  sub condition_.notify_all under lock
(11151|123145408430080)  sub sleep 3 seconds; //?? still holding the lock
(11151|123145436332032) pub50<-delete_participant
(11151|123145438478336) pub54<-delete_participant
(11151|123145408430080)  sub sleep returned //?? release the lock after this
(11151|4347147712) sub condition_.wait returned //?? waiting thread got the lock
(11151|4347147712) <- PublisherService::end
*/

Reply:
Fine, I retract my parenthetical, but my point still stands that it's absolutely fine to notify under the lock. If you have many threads contending for the lock, it's also a potential "pessimization" to release the lock before you notify, since another thread is more likely to have time to "steal" the lock before the notified thread and delay the response of the notifying thread. And you can aggravate _that_ scenario by putting a sleep _before_ the notify.

But the point here wasn't about optimization anyway. My point was about code clarity, and removing the intermediate variable the way I suggested works just fine. The overhead, if any exists, is minimal. If you want to find another way to address the lack of clarity, I'm open to other suggestions.